driver/src/orderbook/streamed/balance.rs

```diff
@@ -0,0 +1,199 @@
+        }
+        if self.withdraw.amount < amount {
+            return Err(Error::WithdrawMoreThanRequested(self.withdraw.amount));
+        }
+        let balance = self.get_balance_internal(batch_id, false)?;
+        let new_balance = balance
+            .checked_sub(amount)
+            .ok_or(Error::WithdrawMoreThanBalance(balance))?;
+        self.balance = new_balance;
+        self.clear_pending_deposit_and_solution_if_possible(batch_id);
+        self.withdraw.amount = 0.into();
+        Ok(())
+    }
+
+    pub fn get_balance(&self, batch_id: BatchId) -> Result<U256, Error> {
+        self.get_balance_internal(batch_id, true)
+    }
+
+    pub fn sell(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.sell_buy_internal(amount, batch_id, Operation::Sell)
+    }
+
+    pub fn buy(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.sell_buy_internal(amount, batch_id, Operation::Buy)
+    }
+
+    pub fn revert_sell(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.revert_sell_buy_internal(amount, batch_id, Operation::Sell)
+    }
+
+    pub fn revert_buy(&mut self, amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        self.revert_sell_buy_internal(amount, batch_id, Operation::Buy)
+    }
+
+    fn get_balance_internal(
+        &self,
+        batch_id: BatchId,
+        include_withdraw: bool,
+    ) -> Result<U256, Error> {
+        let mut balance = self.balance;
+        if self.deposit.batch_id < batch_id {
+            balance = balance
+                .checked_add(self.deposit.amount)
+                .ok_or(Error::MathOverflow)?;
+        }
+        // TODO: this could temporarily fail while not all trades for solution have been received.
+        if self.pending_proceeds.batch_id < batch_id {
+            balance = balance
+                .checked_add(self.pending_proceeds.increase)
+                .ok_or(Error::MathOverflow)?
+                .checked_sub(self.pending_proceeds.decrease)
+                .ok_or(Error::MathOverflow)?;
+        }
+        if include_withdraw && self.withdraw.batch_id < batch_id {
+            // Saturating because withdraw requests can be for amounts larger than balance.
+            balance = balance.saturating_sub(self.withdraw.amount);
+        }
+        Ok(balance)
+    }
+
+    fn clear_pending_deposit_and_solution_if_possible(&mut self, batch_id: BatchId) {
+        if self.deposit.batch_id < batch_id {
+            self.deposit.amount = 0.into();
+        }
+        if self.pending_proceeds.batch_id < batch_id {
+            self.pending_proceeds.increase = 0.into();
+            self.pending_proceeds.decrease = 0.into();
+        }
+    }
+
+    fn sell_buy_internal(
+        &mut self,
+        amount: u128,
+        batch_id: BatchId,
+        operation: Operation,
+    ) -> Result<(), Error> {
+        match self.pending_proceeds.batch_id.cmp(&batch_id) {
```

Comment:
Maybe we can only check if Ordering::Greater in which case we fail. Otherwise we unconditionally update_balance_internal (which would reset the pending proceeds) and then do the get_field + checked_add unconditionally? This would reduce the amount of logic that happens in the match arms to just a simple consistency check.